storage/replica.go

```diff
@@ -2594,6 +2535,81 @@ func (r *Replica) processRaftCommand(
 	return pErr
 }
 
+func (r *Replica) acquireSplitMergeLock(ba roachpb.BatchRequest) func(pErr *roachpb.Error) {
+	arg, ok := ba.GetArg(roachpb.EndTransaction)
+	if !ok {
+		return nil
+	}
+	ict := arg.(*roachpb.EndTransactionRequest).InternalCommitTrigger
+	if unlock := r.acquireSplitLock(ict.GetSplitTrigger()); unlock != nil {
+		return unlock
+	}
+	return r.acquireMergeLock(ict.GetMergeTrigger())
+}
+
+func (r *Replica) acquireSplitLock(split *roachpb.SplitTrigger) func(pErr *roachpb.Error) {
+	if split == nil {
+		return nil
+	}
+
+	rightRng, err := r.store.getOrCreateReplica(split.RightDesc.RangeID, 0, nil)
+	if err != nil {
+		return nil
+	}
+
+	if rightRng.IsInitialized() {
+		// TODO(peter,tschottdorf): TestStoreRangeSplitRaceUninitializedRHS sets
+		// up a situation where one of the stores has already processed the
+		// split. We can see this by noticing that there is an initialized
+		// replica for the right hand side. This seems funky, but the magic in
+		// TestStoreRangeSplitRaceUninitializedRHS is too advanced for me to
+		// deciper. Avoiding locking (and re-initializing) the right hand side
+		// range allows the batch to error when Replica.checkCmdHeader determines
+		// that one of the requests in the batch is trying to operate on keys
+		// outside of the Replica's key range. This results in an "error
+		// executing raft command:" message that can be seen by enabling
+		// -vmodule=replica=1.
```

Comment:
I think we should panic if we find an initialized replica on the RHS while processing a split trigger; that should never happen. If this is happening during TestStoreRangeSplitRaceUninitializedRHS we need to figure out why; it sounds like a regression.